### PR TITLE
Make `00938_fix_rwlock_segfault_long` non flaky

### LINKED ISSUE #70026: `00938_fix_rwlock_segfault_long` test is flaky
https://s3.amazonaws.com/clickhouse-test-reports/69812/7a3217b8c4906b398a796ed3d37d058d9e9bff0f/stateless_tests__ubsan__[1_2].html

```
2024-09-26 06:29:49 Reason: return code:  81
2024-09-26 06:29:49 [f0bf7fb9c24a] 2024.09.26 13:29:49.193130 [ 96089 ] {e3341133-2135-4011-9a1f-2596aaa65297} <Error> executeQuery: Code: 81. DB::Exception: Database test_5dczrcuv does not exist: While executing Columns. (UNKNOWN_DATABASE) (version 24.10.1.411) (from [::1]:47568) (comment: 00938_fix_rwlock_segfault_long.sh) (in query: SELECT name FROM system.tables UNION ALL SELECT name FROM system.columns format Null), Stack trace (when copying this message, always include the lines below):
2024-09-26 06:29:49 
2024-09-26 06:29:49 0. std::exception::capture() @ 0x00000000191222cb
2024-09-26 06:29:49 1. ./contrib/llvm-project/libcxx/include/exception:114: Poco::Exception::Exception(String const&, int) @ 0x00000000385d1705
2024-09-26 06:29:49 2. ./build_docker/./src/Common/Exception.cpp:109: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000002529b72b
2024-09-26 06:29:49 3. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000018fa1136
2024-09-26 06:29:49 4. DB::Exception::Exception<String>(int, FormatStringHelperImpl<std::type_identity<String>::type>, String&&) @ 0x0000000018fba02f
2024-09-26 06:29:49 5. ./build_docker/./src/Interpreters/DatabaseCatalog.cpp:660: DB::DatabaseCatalog::getDatabase(String const&) const @ 0x000000003124d95d
2024-09-26 06:29:49 6. ./build_docker/./src/Storages/StorageMerge.cpp:1382: DB::StorageMerge::DatabaseNameOrRegexp::getDatabaseIterator(String const&, std::shared_ptr<DB::Context const>) const @ 0x00000000339bf21b
2024-09-26 06:29:49 7. ./build_docker/./src/Storages/StorageMerge.cpp:1416: DB::StorageMerge::DatabaseNameOrRegexp::getDatabaseIterators(std::shared_ptr<DB::Context const>) const @ 0x00000000339a0de4
2024-09-26 06:29:49 8. ./build_docker/./src/Storages/StorageMerge.cpp:200: DB::StorageMerge::getColumnSizes() const @ 0x00000000339c14cb
2024-09-26 06:29:49 9. ./build_docker/./src/Storages/System/StorageSystemColumns.cpp:146: DB::ColumnsSource::generate() @ 0x000000002eefdd37
2024-09-26 06:29:49 10. ./build_docker/./src/Processors/ISource.cpp:139: DB::ISource::tryGenerate() @ 0x0000000034cf42b4
2024-09-26 06:29:49 11. ./build_docker/./src/Processors/ISource.cpp:108: DB::ISource::work() @ 0x0000000034cf38da
2024-09-26 06:29:49 12. ./build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:47: DB::ExecutionThreadContext::executeTask() @ 0x0000000034d21e60
2024-09-26 06:29:49 13. ./build_docker/./src/Processors/Executors/PipelineExecutor.cpp:288: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x0000000034d0e5e3
2024-09-26 06:29:49 14. ./build_docker/./src/Processors/Executors/PipelineExecutor.cpp:254: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x0000000034d0ff29
2024-09-26 06:29:49 15. ./contrib/llvm-project/libcxx/include/__functional/function.h:848: ? @ 0x00000000253d2d4f
2024-09-26 06:29:49 16. ./build_docker/./src/Common/ThreadPool.cpp:252: ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() @ 0x00000000253d92e4
2024-09-26 06:29:49 17. ./contrib/llvm-project/libcxx/include/__functional/function.h:848: ? @ 0x00000000253cfc3c
2024-09-26 06:29:49 18. ./build_docker/./src/Common/ThreadPool.cpp:252: void* std::__thread_proxy[abi:v15007]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x00000000253d548a
2024-09-26 06:29:49 19. ? @ 0x00007f75a3bc8ac3
2024-09-26 06:29:49 20. ? @ 0x00007f75a3c5a850
2024-09-26 06:29:49 
2024-09-26 06:29:49 Received exception from server (version 24.10.1):
2024-09-26 06:29:49 Code: 81. DB::Exception: Received from localhost:9000. DB::Exception: Database test_5dczrcuv does not exist: While executing Columns. (UNKNOWN_DATABASE)
2024-09-26 06:29:49 (query: SELECT name FROM system.tables UNION ALL SELECT name FROM system.columns format Null)
2024-09-26 06:29:49 , result:
2024-09-26 06:29:49 
2024-09-26 06:29:49 
2024-09-26 06:29:49 
2024-09-26 06:29:49 stdout:
2024-09-26 06:29:49 
2024-09-26 06:29:49 
2024-09-26 06:29:49 Settings used in the test: --max_insert_threads 2 --group_by_two_level_threshold 1 --group_by_two_level_threshold_bytes 50000000 --distributed_aggregation_memory_efficient 1 --fsync_metadata 0 --output_format_parallel_formatting 1 --input_format_parallel_parsing 0 --min_chunk_bytes_for_parallel_parsing 14304583 --max_read_buffer_size 738582 --prefer_localhost_replica 1 --max_block_size 21515 --max_joined_block_size_rows 44416 --max_threads 3 --optimize_append_index 1 --optimize_if_chain_to_multiif 1 --optimize_if_transform_strings_to_enum 1 --optimize_read_in_order 1 --optimize_or_like_chain 0 --optimize_substitute_columns 0 --enable_multiple_prewhere_read_steps 0 --read_in_order_two_level_merge_threshold 8 --optimize_aggregation_in_order 1 --aggregation_in_order_max_block_bytes 17998632 --use_uncompressed_cache 0 --min_bytes_to_use_direct_io 1 --min_bytes_to_use_mmap_io 10737418240 --local_filesystem_read_method pread_threadpool --remote_filesystem_read_method read --local_filesystem_read_prefetch 1 --filesystem_cache_segments_batch_size 0 --read_from_filesystem_cache_if_exists_otherwise_bypass_cache 0 --throw_on_error_from_cache_on_write_operations 1 --remote_filesystem_read_prefetch 0 --allow_prefetched_read_pool_for_remote_filesystem 0 --filesystem_prefetch_max_memory_usage 64Mi --filesystem_prefetches_limit 10 --filesystem_prefetch_min_bytes_for_single_read_task 1Mi --filesystem_prefetch_step_marks 0 --filesystem_prefetch_step_bytes 0 --compile_aggregate_expressions 1 --compile_sort_description 1 --merge_tree_coarse_index_granularity 17 --optimize_distinct_in_order 1 --max_bytes_before_external_sort 0 --max_bytes_before_external_group_by 10737418240 --max_bytes_before_remerge_sort 2077754751 --min_compress_block_size 2872601 --max_compress_block_size 150089 --merge_tree_compact_parts_min_granules_to_multibuffer_read 21 --optimize_sorting_by_input_stream_properties 1 --http_response_buffer_size 402339 --http_wait_end_of_query True --enable_memory_bound_merging_of_aggregation_results 1 --min_count_to_compile_expression 3 --min_count_to_compile_aggregate_expression 0 --min_count_to_compile_sort_description 3 --session_timezone Africa/Juba --use_page_cache_for_disks_without_file_cache False --page_cache_inject_eviction False --merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability 0.77 --prefer_external_sort_block_bytes 100000000 --cross_join_min_rows_to_compress 1 --cross_join_min_bytes_to_compress 1 --min_external_table_block_size_bytes 0 --max_parsing_threads 10 --optimize_functions_to_subcolumns 0 --parallel_replicas_local_plan 0
2024-09-26 06:29:49 
2024-09-26 06:29:49 MergeTree settings used in test: --ratio_of_defaults_for_sparse_serialization 0.6507869281505183 --prefer_fetch_merged_part_size_threshold 10737418240 --vertical_merge_algorithm_min_rows_to_activate 543572 --vertical_merge_algorithm_min_columns_to_activate 100 --allow_vertical_merges_from_compact_to_wide_parts 0 --min_merge_bytes_to_use_direct_io 8024884581 --index_granularity_bytes 20752379 --merge_max_block_size 16669 --index_granularity 34837 --min_bytes_for_wide_part 0 --marks_compress_block_size 90585 --primary_key_compress_block_size 97011 --replace_long_file_name_to_hash 1 --max_file_name_length 6 --min_bytes_for_full_part_storage 407901778 --compact_parts_max_bytes_to_buffer 157851543 --compact_parts_max_granules_to_buffer 69 --compact_parts_merge_max_bytes_to_prefetch_part 5362788 --cache_populated_by_fetch 0 --concurrent_part_removal_threshold 50 --old_parts_lifetime 40
2024-09-26 06:29:49 
2024-09-26 06:29:49 Database: test_ui7jn4py
```

### PR DESCRIPTION
**Pull request by zvonand**
Fixes #70026

### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit b9f9b50ab74ef1747be893fcb407885a4fd619c7 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70095/b9f9b50ab74ef1747be893fcb407885a4fd619c7/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Builds</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/06a8b8b7f4c8e48a31b4a9564f4abd21018c13fc/builds/report.html">✅ success</a></td></tr>
<tr><td>ClickBench</td><td>Runs [ClickBench](https://github.com/ClickHouse/ClickBench/) with instant-attach table</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/06a8b8b7f4c8e48a31b4a9564f4abd21018c13fc/clickbench__aarch64_.html">✅ success</a></td></tr>
<tr><td>Compatibility check</td><td>Checks that <b>clickhouse</b> binary runs on distributions with old libc versions. If it fails, ask a maintainer for help</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/06a8b8b7f4c8e48a31b4a9564f4abd21018c13fc/compatibility_check__aarch64_.html">✅ success</a></td></tr>
<tr><td>Docker keeper image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/06a8b8b7f4c8e48a31b4a9564f4abd21018c13fc/docker_keeper_image.html">✅ success</a></td></tr>
<tr><td>Docker server image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/06a8b8b7f4c8e48a31b4a9564f4abd21018c13fc/docker_server_image.html">✅ success</a></td></tr>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/924e21f632d22ad26b8380f3286666715d448270/docs_check.html">✅ success</a></td></tr>
<tr><td>Install packages</td><td>Checks that the built packages are installable in a clear environment</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/06a8b8b7f4c8e48a31b4a9564f4abd21018c13fc/install_packages__aarch64_.html">✅ success</a></td></tr>
<tr><td>Integration tests</td><td>The integration tests report. In parenthesis the package type is given, and in square brackets are the optional part/total tests</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/06a8b8b7f4c8e48a31b4a9564f4abd21018c13fc/integration_tests__aarch64__[1_6].html">✅ success</a></td></tr>
<tr><td>Performance Comparison</td><td>Measure changes in query performance. The performance test report is described in detail <a href="https://github.com/ClickHouse/ClickHouse/tree/master/docker/test/performance-comparison#how-to-read-the-report">here</a>. In square brackets are the optional part/total tests</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/06a8b8b7f4c8e48a31b4a9564f4abd21018c13fc/performance_comparison__release__[1_4]/report.html">✅ success</a></td></tr>
<tr><td>Stateful tests</td><td>Runs stateful functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/06a8b8b7f4c8e48a31b4a9564f4abd21018c13fc/stateful_tests__aarch64_.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/06a8b8b7f4c8e48a31b4a9564f4abd21018c13fc/unit_tests__asan_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by alexey-milovidov** [comment]
Ok.

Explanation: querying `system.tables` with a Merge table can throw exception if a database under the Merge table has been dropped at the same time, if you query columns related to total table size in rows/bytes.

File: tests/queries/0_stateless/00938_fix_rwlock_segfault_long.sh [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-# Tags: long, no-debug
+# Tags: long, no-debug, no-parallel
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/00938_fix_rwlock_segfault_long.sh:2]: Missing comments.

Comment (alexey-milovidov) [reply tests/queries/0_stateless/00938_fix_rwlock_segfault_long.sh:2]: Unfortunately, the explanation didn't fit.

Here, we query only the `name` column.
The exception should not happen. Let's fix the underlying cause.

Comment (alexey-milovidov) [reply tests/queries/0_stateless/00938_fix_rwlock_segfault_long.sh:2]: There is no reason the code in the stack trace should be called.

**Comment by alexey-milovidov** [comment]
@zvonand, here is what I meant: #70109

**Pull request closed without merging**